docs/user-guide.md

```diff
@@ -51,19 +51,224 @@ You can find more information about `jelly-cli` in **[its README on GitHub](http
+
+## How does it work – streams
+
+A stream is just a sequence of items. We want to only look at one element of the stream at a time – this limits memory usage and allows us to process infinitely long streams.
+
+The most basic kind of RDF stream is what you would see in an N-Triples or N-Quads file – after all, it's just a sequence of triples:
+
+```turtle
+<https://example.org/jelly> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> _:b1 .
+<https://example.org/jelly> <https://example.org/label> "Jelly" .
+_:b1 <https://example.org/label> "Serialization format"@en .
+```
+
+We call this a *flat RDF stream*. Jelly supports this kind of streams – you will find them useful in transferring large RDF files, query responses, database dumps, and so on.
+
+But, what if our stream contains a sequence of RDF graphs or datasets? For example, if we want to transfer the measurements of a temperature sensor, we would have:
+
+Graph 1:
+
+```turtle
+ex:sensor ex:measuredProperty ex:temperature ;
+    ex:unit ex:Kelvin ;
+    ex:value "280.4"^^xsd:decimal .
+```
+
+Graph 2:
 
-## How to use it – in detail
+```turtle
+ex:sensor ex:measuredProperty ex:temperature ;
+    ex:unit ex:Kelvin ;
+    ex:value "280.6"^^xsd:decimal .
+```
 
-<!-- TODO: the following explanation is unclear. It starts from the perspective of streaming use cases, which is not the most common starting point. Try to reframe it to gently introduce the idea of flat vs grouped streams and go from there. -->
+...and so on. In traditional RDF systems, you had two options. You could modify the data, wrap it in containers, or apply some other processing to make it fit into one graph. Or you could simply put each graph/dataset in a separate file. Jelly offers a simpler way, where multiple graphs or datasets can live within one file. We use *frames* as boundary markers:
 
-Jelly supports several stream types – which stream type you choose depends on your use case (see [stream types](#stream-types) below).
+```turtle
+# Frame 1
+ex:sensor ex:measuredProperty ex:temperature ;
+    ex:unit ex:Kelvin ;
+    ex:value "280.4"^^xsd:decimal .
+
+# Frame 2
+ex:sensor ex:measuredProperty ex:temperature ;
+    ex:unit ex:Kelvin ;
+    ex:value "280.6"^^xsd:decimal .
+```
+
+We call this a *grouped RDF stream*. A Jelly parser can unpack this and process the frames one by one. The best part is that the compression (explained [above](#how-does-it-work--encoding)) is applied **across the entire stream**. So, if an IRI appears in frame 1, and then again in frame 2, we will only have to write it only once. This is very effective for data with repeating patterns, like IoT measurements, nanopublications, encyclopedic entries, or maps (geography).
```

Comment:
```suggestion
We call this a *grouped RDF stream*. A Jelly parser can unpack this and process the frames one by one. The best part is that the compression (explained [above](#how-does-it-work--encoding)) is applied **across the entire stream**. So, if an IRI appears in frame 1, and then again in frame 2, we will only have to write it once. This is very effective for data with repeating patterns, like IoT measurements, nanopublications, encyclopedic entries, or maps (geography).
```

Reply:
let's keep the "only".... it's fine, it's emphasis